src/controllers/handlers/ws-handler.ts

```diff
@@ -3,95 +3,177 @@ import { onRequestEnd, onRequestStart } from '@well-known-components/uws-http-se
+          })
+          metrics.increment('ws_errors', { address: data.address })
+          ws.send(JSON.stringify({ error: 'Error processing message', message: error.message }))
+        }
+      } else if (isNotAuthenticated(data)) {
         try {
           const authChainMessage = textDecoder.decode(message)
-
           const verifyResult = await verify('get', '/', JSON.parse(authChainMessage), {
             fetcher
           })
           const address = normalizeAddress(verifyResult.auth)
 
-          logger.debug('address > ', { address })
+          logger.debug('address > ', { address, clientId: data.clientId })
 
           const eventEmitter = mitt<IUWebSocketEventMap>()
-          changeStage(data, { auth: true, address, eventEmitter, isConnected: true })
+          const transport = await createUWebSocketTransport(ws, eventEmitter, config)
 
-          const transport = UWebSocketTransport(ws, eventEmitter)
+          changeStage(data, {
+            auth: true,
+            address,
+            eventEmitter,
+            isConnected: true,
+            transport
+          })
 
           rpcServer.attachUser({ transport, address })
 
-          transport.on('error', (err) => {
-            if (err && err.message) {
-              logger.error(err)
-            }
+          transport.on('close', () => {
+            rpcServer.detachUser(address)
           })
+
+          if (data.clientId) {
+            wsPool.updateActivity(data.clientId)
+          }
         } catch (error: any) {
-          logger.error(`Error verifying auth chain: ${error.message}`)
+          logger.error(`Error verifying auth chain: ${error.message}`, {
+            clientId: data.clientId
+          })
+          metrics.increment('ws_auth_errors')
           ws.close()
         }
       }
     },
-    close: (ws, code, _message) => {
-      logger.debug(`Websocket closed ${code}`)
+    close: (ws, code, message) => {
       const data = ws.getUserData()
-      if (data.auth) {
-        data.isConnected = false
-        data.eventEmitter.emit('close', code)
+      const { clientId } = data
+      const messageText = textDecoder.decode(message)
+
+      logger.debug('WebSocket closing', {
+        code,
+        message: messageText,
+        clientId,
+        ...(data.auth && { address: data.address })
+      })
+
+      if (data.auth && data.address) {
+        try {
+          data.transport.close()
+          rpcServer.detachUser(data.address)
+          data.eventEmitter.emit('close', { code, reason: messageText })
+          data.eventEmitter.all.clear()
+        } catch (error: any) {
+          logger.error('Error during connection cleanup', {
+            error: error.message,
+            address: data.address,
+            clientId
+          })
+        }
+        metrics.increment('ws_connections', { address: data.address })
+      } else {
+        metrics.increment('ws_connections')
```

Comment:
Maybe these metrics should be triggered by the WS Pool so clients can know nothing about them?